### ISSUE TITLE
Photons interacting with the same resonance in small volume

### ISSUE THREAD

**Issue by saultyevil** [sirocco-rt/sirocco#856]
Changes introduced to stop a photon from interacting with the same resonance in a small volume have been added to Python.  When this happens, the photon is reset as an inwind photon and it doesn't scatter with the resonance.

In the case of some optically thick TDE models with large NPHOT, this is happening enough times to cause the code to exit.

The current volume is a sphere of radius 1e5 cm from the last resonance scatter of the photon. The cell specific dfudge in this model ~6e5, so I think in most cases the step size in `reposition()` is fine, but could also possibly not be a big enough step to get away. However, if the photon is close to hitting the star, then I think the step size is smaller than dfudge which could be < 1e5. It could be photons like this triggering it.

Currently adding more diagnostics to find out where exactly this is happening and how big certain distances are. Also searching for a short test case for a single processor...

**Comment by saultyevil** [comment]
Looking at a handful of my 1D TDE models, the error related to repeated resonance scatters are appearing to happen in similar places between models. This region is where the temperature of the model starts to take a dive, also where the wind is struggling to converge or we're having numerical errors.

The repeated resonance error doesn't happen once for a photon/interaction, but happens upwards of 25 times before the photon manages to avoid creating the error with the specific resonance. If you haven't seen the error before, it looks like this:

```
Error: translate_in_wind: nres  2837 repeat after motion of  3.046e+04 of phot 487069 in ion cycle 26 spec cycle  0 in plasma 34 stat(0 -> 1)
```  

I've also output how far the photon was previously repositioned,

```
translate_in_wind: one->dfudge 6.257783e+05 DFUDGE 6.257783e+05 reposition_ds 6.257783e+05
(wind cell dfudge, global dfudge, reposition distance)
```

So the photon scatters with the resonance and is repositioned by 6e5 cm. Then on the next translation, the photon moves < 1e5 cm and wants to interact with the same resonance. I don't think we're in the regime where we have to worry about Soblev either, the wind is still 0.1c.

Here's the electron density for one of the models.. which isn't any higher than what we simulated in the past TDE models. The only difference is that it's 1D, so photons can't go around the problem area.

![ne_model](https://user-images.githubusercontent.com/19627279/117631057-c20e3b00-b173-11eb-99b4-e6949df780e9.png)

**Comment by saultyevil** [comment]
Do you have any example errors form Keith's model, i.e. the distance photons move between interactions when the error is triggered?

What I've found is that most photons are moving ~ 5e4cm before interacting again. I think this means either 1) the distance reposition moves photons is not large enough or 2) the distance to check for repeated interactions is too large.

I do not have a short model for debugging, just yet. I am trying to get a model with the RNG state dumped out to try and re-create the problem but even then it's still looking to be expensive.

**Comment by saultyevil** [comment]
I believe this error has been very difficult to track down because we are accidentally suppressing them.

Here's how I think this is happening. In `tanslate_in_wind()` we have this variable `translate_in_wind_res_count` which is incremented after we move the photon. This happens every call to `translate_in_wind()`. I believe It is used to suppress the number of errors we see being triggered. Specifically, it is used in the if statement to catch repeated resonance scattering such that the error is only caught if `translate_in_wind_res_count < 5000`.

```c
move_phot (p, ds_current);
translate_in_wind_res_count++;

if (*nres > -1 && *nres <= NLINES && *nres == p->nres && istat == P_SCAT && translate_in_wind_res_count < 5000)
{   
   if (ds_current < 1e5)
   {
    ...
   }
}
```

If you remove the `translate_in_wind_res_count < 5000` part of the statement, then I have attached a model which displays the error messages very early on in the first ionization cycle.

[input.zip](https://github.com/agnwinds/python/files/6523411/input.zip)

I can't say for certain why this error triggered only in long runs, but my current running theory is that we were seeing a signed integer overflow. This happens when a signed integer is incremented past its max value (2,147,483,647), which I think makes sense as I'm using 1e8 photons each scattering 1000s of times.

Reaching the max value is classified as undefined behaviour, but I believe most machines will turn the int into a negative number so ` translate_in_wind_res_count < 5000` is triggered.

**Comment by kslong** [comment]
@saultyevil   The current version of dev does not produce any errors on the .pf file in input.zip file.  Do you know the reason for this.  Assuming this statement is correct, Is there a new example file for this problem.  Is the agn_macro_gamma.

**Comment by saultyevil** [comment]
You need to remove the `translate_in_wind_res_count < 5000` part in the check for repeated resonance scattering in `translate_in_wind()` as I haven't uploaded any changes to remove that part in dev. Without that change the error isn't able to happen after 5000 calls to `translate_in_wind()` until the integer `translate_in_wind_res_count` overflows, then the error message can trigger again.

You can then reproduce the error using parameter file in input.zip, as well as a number of our standard models such as cv_standard.

**Comment by saultyevil** [comment]
The situation here has been improved dramatically, and has been fixed by the following changes:

* Photons are no longer allowed to interact with the same resonance within a spherical volume with a radius equal to the cell specific dfudge. Photons will now either not interact, or interact with a different line or with the continuum.
* We no longer reposition a photon when it resonance scatters and we have removed the function completely.

There were (small?) differences to some of the regression tests, mainly agn_macro_gamma. This was caused by a slight difference in ionization state due to different photon paths. Some other models were affected by noise. However, the most recent regression tests (after some code clean up) showed that these differences are now even smaller.
